juju/juju
# [JUJU-1514] Move existing package doc comments to doc.go and tidy up

In light of wanting to use {package}/doc.go consistently, I thought
it'd be good to update all our existing package doc comments to use
doc.go. There were 33 doc.go's previously, now there are 95 (that is,
all the packages that have a package doc comment).

I also made things a bit more consistent, for example:

* Starting the package doc comment with "Package {x} ...". We did this
  most of the time, but not always.
* Avoiding duplicate package comments in multiple files, for example
  the ./api/client/charms package.
* Removing the copyright message from a few package doc comments (by
  adding a blank line between the copyright notice and the "package"
  directive so it wasn't included in the package comment).

I wrote a little Python script (see below) to find all the package doc
comments and enable me to diff the result, to ensure I didn't stuff
anything up. [**Diff here**](https://gist.github.com/benhoyt/7b47c3a691e716460a555909d1020469/revisions) - the first 1/3 or so of that page shows the actual
diff, the rest shows the original output.

## Checklist


## QA steps

I used a small Python script to find all the package doc comments, and
generate the before/after:

```python
import os
import subprocess

packages = []
for root, dirs, files in os.walk('.'):
	has_go = any(f.endswith('.go') for f in files)
	has_doc = any(f == 'doc.go' for f in files)
	if has_go:
		packages.append((root, has_doc))

packages.sort()
for package, has_doc in packages:
	process = subprocess.run(['go', 'doc', package], capture_output=True, check=True)
	output = process.stdout.decode('utf-8')

	short = []
	for line in output.splitlines():
		if line.startswith(('func ', 'var ', 'type ', 'const ')):
			break
		short.append(line)
	short = '\n'.join(short)
	short = short.strip()

	if not short:
		continue
	if '\n' not in short:
		continue
	if 'is a generated GoMock package' in short:
		continue

	contains = 'Package '
	if contains not in short:
		raise Exception(f'Package {package} does not contain {contains!r}: {short!r}')

	# print('// '+'='*60, 'has doc.go' if has_doc else 'no doc.go')
	print('// '+'='*60)
	print(short)
	print()
```
